diff --git a/mlir/cmake/modules/AddMLIRPython.cmake b/mlir/cmake/modules/AddMLIRPython.cmake
index 7d47fd0bca..52bbf7d64b 100644
--- a/mlir/cmake/modules/AddMLIRPython.cmake
+++ b/mlir/cmake/modules/AddMLIRPython.cmake
@@ -372,22 +372,24 @@ function(build_nanobind_lib)
   )
 
   if(NOT MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES
-     AND (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL))
+     AND (MSVC OR LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL))
     # Avoid some warnings from upstream nanobind.
     # If a superproject set MLIR_DISABLE_CONFIGURE_PYTHON_DEV_PACKAGES, let
     # the super project handle compile options as it wishes.
     _mlir_python_eh_rtti_flags()
-    target_compile_options(${NB_LIBRARY_TARGET_NAME}
-      PRIVATE
-        -Wno-c++98-compat-extra-semi
-        -Wno-cast-qual
-        -Wno-covered-switch-default
-        -Wno-deprecated-literal-operator
-        -Wno-nested-anon-types
-        -Wno-unused-parameter
-        -Wno-zero-length-array
-        -Wno-missing-field-initializers
-        ${eh_rtti_enable})
+    if(NOT MSVC)
+      target_compile_options(${NB_LIBRARY_TARGET_NAME}
+        PRIVATE
+          -Wno-c++98-compat-extra-semi
+          -Wno-cast-qual
+          -Wno-covered-switch-default
+          -Wno-deprecated-literal-operator
+          -Wno-nested-anon-types
+          -Wno-unused-parameter
+          -Wno-zero-length-array
+          -Wno-missing-field-initializers)
+    endif()
+    target_compile_options(${NB_LIBRARY_TARGET_NAME} PRIVATE ${eh_rtti_enable})
   endif()
 
   # Apply caller-provided extra options last so they have higher precedence.
