diff --git a/mlir/include/mlir/Dialect/LLVMIR/ROCDLDialect.td b/mlir/include/mlir/Dialect/LLVMIR/ROCDLDialect.td
index 0807cf8cf04a..eca64b9a3455 100644
--- a/mlir/include/mlir/Dialect/LLVMIR/ROCDLDialect.td
+++ b/mlir/include/mlir/Dialect/LLVMIR/ROCDLDialect.td
@@ -149,7 +149,7 @@ def ROCDL_Dialect : Dialect {
      "::mlir::DenseI32ArrayAttr":$reqd_work_group_size,
      "::mlir::StringAttr":$flat_work_group_size,
      "::mlir::IntegerAttr":$max_flat_work_group_size,
-     "::mlir::IntegerAttr":$waves_per_eu,
+     "::mlir::StringAttr":$waves_per_eu,
      "::mlir::BoolAttr":$unsafe_fp_atomics,
      // Correspond to LLVM metadata of the same name
      "::mlir::UnitAttr":$last_use,
diff --git a/mlir/lib/Target/LLVMIR/Dialect/ROCDL/ROCDLToLLVMIRTranslation.cpp b/mlir/lib/Target/LLVMIR/Dialect/ROCDL/ROCDLToLLVMIRTranslation.cpp
index 42393218f870..4ecf70895442 100644
--- a/mlir/lib/Target/LLVMIR/Dialect/ROCDL/ROCDLToLLVMIRTranslation.cpp
+++ b/mlir/lib/Target/LLVMIR/Dialect/ROCDL/ROCDLToLLVMIRTranslation.cpp
@@ -103,16 +103,19 @@ public:
       if (!func)
         return op->emitOpError(Twine(attribute.getName()) +
                                " is only supported on `llvm.func` operations");
-      auto value = dyn_cast<IntegerAttr>(attribute.getValue());
-      if (!value)
-        return op->emitOpError(Twine(attribute.getName()) +
-                               " must be an integer");
 
       llvm::Function *llvmFunc =
           moduleTranslation.lookupFunction(func.getName());
       llvm::SmallString<8> llvmAttrValue;
-      llvm::raw_svector_ostream attrValueStream(llvmAttrValue);
-      attrValueStream << value.getInt();
+      if (auto strValue = dyn_cast<StringAttr>(attribute.getValue())) {
+        llvmAttrValue.append(strValue.getValue());
+      } else if (auto intValue = dyn_cast<IntegerAttr>(attribute.getValue())) {
+        llvm::raw_svector_ostream attrValueStream(llvmAttrValue);
+        attrValueStream << intValue.getInt();
+      } else {
+        return op->emitOpError(Twine(attribute.getName()) +
+                               " must be a string or an integer");
+      }
       llvmFunc->addFnAttr("amdgpu-waves-per-eu", llvmAttrValue);
     }
     if (dialect->getFlatWorkGroupSizeAttrHelper().getName() ==
diff --git a/mlir/test/Target/LLVMIR/rocdl.mlir b/mlir/test/Target/LLVMIR/rocdl.mlir
index 9b767d857447..20a4900e1985 100644
--- a/mlir/test/Target/LLVMIR/rocdl.mlir
+++ b/mlir/test/Target/LLVMIR/rocdl.mlir
@@ -150,6 +150,13 @@ llvm.func @kernel_func_waves_per_eu()
   llvm.return
 }
 
+llvm.func @kernel_func_waves_per_eu_range()
+    attributes {rocdl.kernel, rocdl.waves_per_eu = "1,1"} {
+  // CHECK-LABEL: amdgpu_kernel void @kernel_func_waves_per_eu_range()
+  // CHECK: #[[$KERNEL_WAVES_PER_EU_RANGE_ATTR:[0-9]+]]
+  llvm.return
+}
+
 llvm.func @kernel_func_unsafe_fp_atomics()
     attributes {rocdl.kernel, rocdl.unsafe_fp_atomics = true} {
   // CHECK-LABEL: amdgpu_kernel void @kernel_func_unsafe_fp_atomics()
@@ -2241,4 +2248,5 @@ llvm.func @rocdl_dot_fp8_family(%i32: i32, %f32: f32) -> f32 {
 // CHECK-DAG: ![[$REQD_WORK_GROUP_SIZE]] = !{i32 16, i32 4, i32 2}
 // CHECK-DAG: ![[$REQD_BLOCK_SIZE]] = !{i32 8, i32 4, i32 2}
 // CHECK-DAG: attributes #[[$KERNEL_WAVES_PER_EU_ATTR]] = { "amdgpu-flat-work-group-size"="1,256" "amdgpu-waves-per-eu"="2" "uniform-work-group-size" }
+// CHECK-DAG: attributes #[[$KERNEL_WAVES_PER_EU_RANGE_ATTR]] = { "amdgpu-flat-work-group-size"="1,256" "amdgpu-waves-per-eu"="1,1" "uniform-work-group-size" }
 // CHECK-DAG: attributes #[[$KERNEL_UNSAFE_FP_ATOMICS_ATTR]] = { "amdgpu-flat-work-group-size"="1,256" "amdgpu-unsafe-fp-atomics"="true" "uniform-work-group-size" }
